### ISSUE TITLE
Check dev experience on starting an issue

### ISSUE THREAD

**Issue by rndquu** [ubiquity-os/plugins-wishlist#26]
There are certain kinds of tasks which must be completed only by experienced developers.

Check the following issues for example:
- https://github.com/ubiquity/ubiquity-dollar/issues/927
- https://github.com/ubiquity/ubiquity-dollar/issues/563

It would be great to know that collaborator solving the above issues has prior experience with solidity and ethereum.

Possible solution could be to use chatgpt to parse collaborator's github or CV to make sure a contributor is experienced enough to `/start` an issue.

**Comment by 0x4007** [comment]
>  github or CV

Toss out CVs I never believe them. GitHub though, there is proof of work. I am a big fan of using GitHub for this purpose. 

> make sure a contributor is experienced enough to /start an issue.

How can you determine this? 
1. Amount of repositories containing Solidity. 
2. Amount of commits containing Solidity lines of code.
3. Amount of Solidity files committed. 
4. Amount of contracts deployed from their address on chain.

**Comment by 0x4007** [comment]
People lie on CVs all the time. It's useless data compared to a portfolio of work.

**Comment by gentlementlegen** [comment]
I think within a CV you can write anything. GitHub is reliable, maybe we could try using the statistics? https://docs.github.com/en/rest/metrics/statistics?apiVersion=2022-11-28
I also think that LinkedIn could be a way, but this will require users to have a LinkedIn and agree to share that data.

**Comment by Keyrxng** [comment]
We could leverage [Gitroll](https://gitroll.io/scan
), there is a commercial [version](https://gitroll.io/for-companies) too, which creates a CV based on the their github history.

[My profile scan](https://gitroll.io/profile/uOp67oGeYgBNu5MjHSCmHHoqY0qV2)

That was a public scan which does not include private contributions, you can log in and do a personal one for more info.

It's not very fast in completing scans although once a scan has been completed you can search for it. So first see if a scan exists, if not fire one (scanning my profile takes between 3-5 minutes)

<hr/>

https://github-readme-stats.vercel.app/api/top-langs/?username=keyrxng

We could fetch top language stats from this endpoint and parse it for solidity percentage. Seems brittle but you'd expect anyone with a relative amount of pushed solidity code would have more than 15/20% I think

**Comment by gentlementlegen** [comment]
Didn't know about Gitroll, it's pretty fun. The risk is that it takes into account forks and I know some people just keep forking repos for some reason, I don't know if it takes into account the lines that you actually wrote. But that could be a starting point which might save tons of development hours.

**Comment by Keyrxng** [comment]
> Didn't know about Gitroll, it's pretty fun.

Surprised you missed it, it had 15 minutes of fame on twitter and everyone was doing it but it was some time ago.

> The risk is that it takes into account forks and I know some people just keep forking repos for some reason, I don't know if it takes into account the lines that you actually wrote. 

https://gitroll.io/profile/uOp67oGeYgBNu5MjHSCmHHoqY0qV2/repos

> (35 detected · 10 scannable)

As far as I can tell it detects forks but does not count them as code you've authored. The repos you see in that link are only my repos so I think it only accounts for code you've authored in your own repos.

> But that could be a starting point which might save tons of development hours.

But yeah that's why I mentioned it I doubt it suits our needs out of the box but something to work towards maybe

**Comment by Keyrxng** [comment]
Well either way,

> download the source code and push as their own, suddenly they are a 10x

we only have available the contributor's public contributions and repos (unless, does the bot have private repo access of a user via the authToken?) which is susceptible to the same sort of manipulation.

<hr/>

1. Is this XP-Gating specifically for blockchain/web3 tasks, or is this the official XP-Gate plugin?
2. Why isn't this an extension of the `/start|stop` command and is being built as a standalone plugin?
3. Later then would this be refactored to use the real XP system based on earnings for XP-gating or is that another plugin?

**Comment by Keyrxng** [comment]
@gentlementlegen can you price this with a few hours I've spent a couple on it so far and I have a couple more expanding the test suite and optimizing/refining things

**Comment by 0x4007** [comment]
> @gentlementlegen can you price this with a few hours I've spent a couple on it so far and I have a couple more expanding the test suite and optimizing/refining things

Seems like this is roughly a day task

**Comment by 0x4007** [comment]
@rndquu do you have any remarks on your vision and how granular the guarding will be? Ref: https://github.com/ubiquibot/command-start-stop/pull/17#issuecomment-2254365866

I think labeling is nice to have because even our [devpool.directory](https://devpool.directory/) supports this right now. You can see a "UI/UX" and "Research" task.

I suppose that in the config, the partner should be able to associate an arbitrary label name to a type of code for the check. This is obviously restricted to coding tasks, but seems straightforward to implement if GitHub recognizes code types for the statistics. 

<img width="762" alt="Screenshot 2024-07-29 at 16 17 35" src="https://github.com/user-attachments/assets/3ec5e6a2-7db3-4152-b9d6-9e2105d26bf6">

**Comment by rndquu** [comment]
> @rndquu do you have any remarks on your vision and how granular the guarding will be? Ref: https://github.com/ubiquibot/command-start-stop/pull/17#issuecomment-2254365866

Granular enough to reduce the number of false positives.

Not sure how @Keyrxng is going to implement this plugin but at first glance we could:
1. Fetch original issue repository languages via https://docs.github.com/en/rest/repos/repos?apiVersion=2022-11-28#list-repository-languages (this eliminates the need to issue labels)
2. Fetch languages from repositories where a contributor once committed
3. Match either using gpt either some heuristics that makes sense

<img width="1512" alt="Screenshot 2024-07-30 at 20 10 28" src="https://github.com/user-attachments/assets/922fe575-3d74-42ea-ac69-f9f2c4fb0b6b">

I don't think the 1st version of this plugin must be super accurate. But it must be accurate enough to allow assigning [this](https://github.com/ubiquity/ubiquity-dollar/issues/926) issue only to somebody with solidity experience.

**Comment by Keyrxng** [comment]
@rndquu the open [pr](https://github.com/ubiquibot/command-start-stop/pull/17) implements things similarly but in reverse and meets the requirements of the spec, it just needs refined a little.

1. Scan user code statistics (as seen in my readme) not the repo
2. Have repo config defined item `mostImportantLanguage: {Solidity: 25}` with a threshold. Since V1 aims to restrict Solidity tasks primarily this works fine as only a couple of solidity-heavy repos exist.

This approach gives us configurability at the repo level but not the org level although a default baseline could be set, `age > 1 year && ts > 5 etc...`

---

The reason I chose not to use the repo code stats for V1 is because of [situations](https://github.com/ubiquity/keygen.ubq.fi) like in the screenshot below. It's not a problem for the solidity repos but will be a problem in others and will need to be addressed specifically as off the top of my head idk how we'd handle this elegantly

![image](https://github.com/user-attachments/assets/72cbe81c-a03f-474e-b109-da3a46b266b6)

---

Overall, I think the current PR meets the requirements for V1. V2 will likely revolve around the real XP system which would follow the same mapped config setup as labels/tags probably so it should be thought out and tasked I think.

**Comment by gentlementlegen** [comment]
@Keyrxng What unit is it exactly  when you say `Solidity: 25` for example?

**Comment by Keyrxng** [comment]
> @Keyrxng What unit is it exactly when you say `Solidity: 25` for example?

1-100, maybe it can be made clearer what sort of threshold it is.

rndquu's language stats looks like:
![image](https://github.com/user-attachments/assets/d9c4a278-d2d5-4efb-87c4-47f2ab7a036d)

molecula451's looks like:
![image](https://github.com/user-attachments/assets/b6110d6c-e72d-4c01-bcf3-7dee92f9997d)

gitcoindev's looks like:
![image](https://github.com/user-attachments/assets/198d987d-1579-489e-bbd9-a15756d3dbcf)

0x4007's looks like:
![image](https://github.com/user-attachments/assets/55c48b2b-1212-4d69-803c-6f325ab1098f)

Mine looks like:
![image](https://github.com/user-attachments/assets/48a178bd-70ab-4b3c-ab7b-9c2832eff937)

---

I had only looked at mine and @rndquu's stats in QA before but looking over more it's making me doubtful that this is going to be as effective as I first thought.

Any thoughts?

**Comment by gentlementlegen** [comment]
This seems to be the percentage of a certain language you are mostly using through your repos isn't it? Which means a beginner that only did one project in TypeScript would get a TypeScript: 100`? Or did I misunderstand those numbers.

**Comment by Keyrxng** [comment]
> This seems to be the percentage of a certain language you are mostly using through your repos isn't it? Which means a beginner that only did one project in TypeScript would get a `TypeScript: 100`? Or did I misunderstand those numbers.

No that sounds right afaik and is why I also included the other markers at first as they'd help out balance that scenario and others that are similar. It looks like this will need to do at least a little bit of manual validation on the user's PRs/repos anyway but without a concrete XP based system we are up against it as there are lots of ways to spoof your github stats and data.

**Comment by 0x4007** [comment]
I think we should get all of the user's commits and determine which languages they are committing in.

**Comment by Keyrxng** [comment]
> We can do a ton of requests because we have a six hour runtime.

not really ideal for users waiting any length of time for a response plus we are bound by the fact that `command-start-stop` is a worker plugin not an action plugin

The endpoint I'm using to gather user stats is open source and self-hostable if we want to go down that route and keep the plugin fast

**Comment by 0x4007** [comment]
> > We can do a ton of requests because we have a six hour runtime.
> 
> not really ideal for users waiting any length of time for a response plus we are bound by the fact that `command-start-stop` is a worker plugin not an action plugin

Cache might be useful, then we only need to run it once per user. We can rerun if they previously were not high enough level? But yes I agree any lag is not attractive.

**Comment by Keyrxng** [comment]
> Cache might be useful, then we only need to run it once per user. We can rerun if they previously were not high enough level?

I thought of this and we might be able to get away with it with one user with the worker, but if it's a team then that's potentially tens of thousands of commits. Making assumptions here obviously and I'll only know after testing but I expect it to be problematic.

The little work I've done on the faucet, I read that worker limits while appear to be time based are more memory-based than anything else. I have less exp here than any of you folks obviously but if that is your opinion also then it becomes a little more problematic

> I don't think this is a good idea to maintain all this infra for this plugin.

agreed it's not ideal for just this alone so I will proceed with other suggestions

**Comment by rndquu** [comment]
> > @Keyrxng What unit is it exactly when you say `Solidity: 25` for example?
> 
> 1-100, maybe it can be made clearer what sort of threshold it is.
> 
> rndquu's language stats looks like: ![image](https://private-user-images.githubusercontent.com/106303466/353711370-d9c4a278-d2d5-4efb-87c4-47f2ab7a036d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI0NTg5MjksIm5iZiI6MTcyMjQ1ODYyOSwicGF0aCI6Ii8xMDYzMDM0NjYvMzUzNzExMzcwLWQ5YzRhMjc4LWQyZDUtNGVmYi04N2M0LTQ3ZjJhYjdhMDM2ZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzMxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDczMVQyMDQzNDlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mZDcxNTg1MmIwZWE0NzQ5ZDQ1YTM5MWE0OWVjYTgyMjk3MTk0MTdlM2FiZGQzOWY5OWFiZDE4ODdkMGI0YmFlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.F38LM9PjEbhXJZAibjHY5dHpBgte9j7OQMVnMf6_RV0)

As far as I understand those stats are taken from commits, not from forked repositories so it requires quite an effort to spoof those stats. 

Anyway I think it's enough for v1. Setting a label like `Solidity: 10` for https://github.com/ubiquity/ubiquity-dollar/issues/927 should be enough for initial pre-screen.

**Comment by 0x4007** [comment]
@Keyrxng this can run async in the background via the GitHub action runner. Here is a user flow:

1. User self assigns via /start
2. Bot assigns them 
3. Bot runs a "background check"
4. A few minutes later the bot unassigns with a descriptive warning message "not enough xp" etc

**Comment by Keyrxng** [comment]
poor guy probs gets tagged thousands of times per week 😂

> Then the action runner can check all their commits at its leisure.

So should `command-start-stop` be both a worker (for the initial response) then it should dispatch to the workflow within the same repo?

Or should it be a separate plugin which runs after and has it's own config etc?

**Comment by Keyrxng** [comment]
I just had a thought based on [this issue](https://github.com/ubiquity/pay.ubq.fi/issues/267) that we should also build in a `core-team` or similar check so we can restrict issues that seem would be a good fit for it

@0x4007 It would be easy enough to build into the open PR if you could define label schema

- `Core-team: (biz-dev)` and it's restricted to teams as an idea

**Comment by 0x4007** [comment]
Separate plugin. It's literally a couple of minutes max of wasted effort. I think this is acceptable. Can post a warning while it works. 

Private repo is sufficient.

**Comment by 0x4007** [comment]
> @Keyrxng this can run async in the background via the GitHub action runner. Here is a user flow:
> 
> 1. User self assigns via /start
> 2. Bot assigns them 
> 3. Bot runs a "background check"
> 4. A few minutes later the bot unassigns with a descriptive warning message that tags them "@user not enough xp" etc
> 
> Then the action runner can check all their commits at its leisure. 

I just realized with this plugin enabled we should reply with...

```diff
# Analyzing your work to see if you qualify for this task
```

...comment before assigning. If they pass, then assign. If not, then edit the message saying that they require more experience. Perhaps something like

```diff
! You need more TypeScript projects on your GitHub profile in order to be eligible for this task. 
```

**Comment by Keyrxng** [comment]
> I just realized with this plugin enabled we should reply with...

Review has taken me in the direction of this running async after `/start`. So with this change we'd need to have it run on the same command as in the `start-stop` manifest and in parallel to `/start`, we'd then update the comment during the `issues.assigned` event. 

However, if the self-assign checks fail then that event won't fire and so would we delete the comment from within the `/start` logic or add a listener for the error comment from `/start` and then remove the comment in this plugin?

> If they pass, then assign. If not, then edit the message saying that they require more experience. Perhaps something like

So this should run before `/start` and it should pass the results of each xp-check for each user into `/start` so it knows whether to assign them or not?

> It could also be really interesting to include a gif of a loader spinner for some of these transient comments.

That would be cool, why not task it out and make an on-brand logo loader?

**Comment by 0x4007** [comment]
![image](https://github.com/user-attachments/assets/6ba235ed-ee4e-4aad-8f3e-20884e5b84bc)

Testing if it works

**Comment by Keyrxng** [comment]
Looks cool sort of Tron vibes and you can size it with an img ele

<img src="https://github.com/user-attachments/assets/6ba235ed-ee4e-4aad-8f3e-20884e5b84bc" width="300" height="200" alt="image">

**Comment by Keyrxng** [comment]
Looks like the repo for this was deleted maybe as I cannot find it lmao https://github.com/ubiquibot/task-xp-guard/pull/1

**Comment by 0x4007** [comment]
You make your own repo. On your own org. We copy it when it's finished.

**Comment by Keyrxng** [comment]
The most recent PR was deleted so I can't reference the conversation so I want to clarify and summarize here as I am currently refactoring my working approach to use your commit counting strategy.

In short, I should collect a user' commits and then reduce the commits down to a tally of file extensions that appeared once per commit regardless of what that commit contains all that matters is appearing in the commit. "deletions" should be ignored from the tally.

1. I need to collect the user' repos. I assume use GraphQL with something like- 
```gql
query userRepositories($login: String!) {
            user(login: $login) {
                repositories(first: 100, ownerAffiliations: [OWNER, COLLABORATOR], isFork: false) {
                    nodes {
                        name
                        url
                    }
                }
            }
        }

```
3. Now that I have repos I can use `octokit.paginate(octokit.repos.listCommits` to collect the all of the commits from that repo or GraphQL to obtain the linear commit history.
4. I can filter out all non-assignee commits and then I need to use `octokit.paginate(octokit.repos.getCommit` on each commit, as `listCommits` is a thin payload and doesn't contain the things we need to parse.
5. I accumulate from the `getCommit` responses and the end result is the user experience.

I'm at the accumulation stage right now and using @0x4007 as the account to test with.

```
Total commits fetched:  1261
Covering repos:  31
Fetching commit details for 0x4007/nv
Fetching commit details for RameshPR/VR01B
Fetching commit details for 0x4007/nv2
Fetching commit details for 0x4007/proxy-handler
Fetching commit details for 0x4007/google-apps-scripts-general
Fetching commit details for 0x4007/nvcrm
Fetching commit details for 0x4007/lrdscx-stripe
Fetching commit details for 0x4007/begin-app
Fetching commit details for 0xWildhare/SBTToken
Fetching commit details for 0x4007/uad-ui-archived
Fetching commit details for 0x4007/uniswap-lp-apr
Fetching commit details for 0x4007/opensea-cashback
Fetching commit details for 0x4007/hardhat-typescript-starter
Fetching commit details for Steveantor/msgtester
Fetching commit details for ubiquity/github-labels-processor
Fetching commit details for 0x4007/0x4007
Fetching commit details for 0x4007/scraper-parent-test
Fetching commit details for 0x4007/metamask-cache-test
Fetching commit details for 0x4007/parcel-ethers-test
Fetching commit details for 0x4007/the-grid
Fetching commit details for 0x4007/ubiquibot-sandbox
Fetching commit details for ubiquity-os/ubiquity-os-kernel
Fetching commit details for 0x4007/blog.pavlovcik.com
Fetching commit details for 0x4007/link-pulls
Fetching commit details for 0x4007/hue
Fetching commit details for 0x4007/clone-all-repositories
Fetching commit details for ubiquity-os-marketplace/conversation-rewards
Fetching commit details for 0x4007/cx-linkedin-scraper
Fetching commit details for powerhouse-inc/powerhouse-mirror
Fetching commit details for 0x4007/pull-time-estimator
Fetching commit details for ubiquity-os/ubiquity-os-plugin-installer
Experience via commits:  {
  js: 166,
  md: 176,
  htaccess: 1,
  conf: 2,
  array: 3,
  html: 140,
  htm: 50,
  css: 293,
  json: 1197,
  LICENSE: 5,
  php: 6,
  'service/error_log': 2,
  error_log: 2,
  ts: 2858,
  gitignore: 95,
  map: 19,
  lock: 174,
  npmignore: 3,
  eslintrc: 70,
  MD: 2,
  sh: 65,
  png: 859,
  svg: 2,
  txt: 18,
  tsbuildinfo: 3,
  contracts: 11,
  tsx: 305,
  mp4: 6,
  DS_Store: 16,
  eot: 26,
  'eot?': 10,
  ttf: 30,
  woff: 26,
  gitmodules: 7,
  prettierignore: 16,
  eslintignore: 6,
  yml: 349,
  log: 58,
  prettierrc: 41,
  eslintcache: 3,
  babelrc: 6,
  example: 19,
  solhintignore: 1,
  sol: 1,
  nvmrc: 20,
  commitlintrc: 10,
  cache: 2,
  csv: 2,
  SingletonLock: 3,
  'Default/Extension Scripts/LOG': 2,
  old: 77,
  ldb: 7,
  db: 15,
  DevToolsActivePort: 2,
  SingletonCookie: 2,
  SingletonSocket: 2,
  Variations: 2,
  RunningChromeVersion: 1,
  pma: 3,
  'cache/ShaderCache/data_3': 1,
  lockb: 10,
  toml: 5,
  cjs: 9,
  mjs: 4,
  editorconfig: 2,
  ico: 5,
  license: 2,
  jpg: 2,
  mkv: 2,
  gitconfig: 1,
  'github/CODEOWNERS': 6,
  parcelrc: 1,
  terserrc: 1
}
Time taken:  319540.2787
Total getCommit API calls made:  1261
```
I've removed hundreds of bad entries manually (left a couple in) and will need to add handling for it but assuming all three aspects are acceptable. 1. Repo collection 2. Tally output 3. Time taken. then how are we transforming these numbers into stats? Just determining weight across the number of different languages?

In a team scenario the call count and time taken could easily 2-3x and so it's eating a huge chunk of our rate limit every time a task is started and the delay is ~5 mins per contributor. If we try to make things go faster we'll hit the secondary rate limit which will knock out all other plugins. 

I think a cache was mentioned but I'm not sure how that would work, when would we update a cache entry - every n tasks?

GraphQL does not expose what we need or it does but not the way we need it, i.e it exposes the tree with all files and we can pinpoint files with a path via the tree at that point in time but it doesn't contain the files in the commit. I've been using the [explorer](https://docs.github.com/en/graphql/overview/explorer) for introspection manually and GPT both telling me it's not possible, maybe we are both wrong but I don't think this is a feasible approach without removing the calls to `getCommit`.

I tried to use Blame and Contribution Graph data from GraphQL too but all roads lead nowhere.

**Comment by 0x4007** [comment]
Maybe we can optimize the rate limits in a different way. For now we can let the job run slowly and as for cache we just need to store the totals somewhere. 

The first run will be the heaviest, maybe a separate app can handle that. Later runs could "sync" from the last time it ran. For example since you just checked today, and if I start a task tomorrow, it will only check my last one day of commits which should be a relatively tiny amount. 

Seems like the stats you aggregated look pretty great. It shows I'm pretty experienced with typescript which is what I expected.

**Comment by Keyrxng** [comment]
I'm sorry I do not understand how that approach is better/more effective than what I had implemented, if the primary goal is detecting the languages which appears most across a user' commits across repos they own and have collaborated on. My approach indexed the same repos and obtained the same final result (that you know TS) in under 10s and used <10 api calls.

[My comment from before](https://github.com/ubiquity-os/plugins-wishlist/issues/26#issuecomment-2259741041):
![image](https://github.com/user-attachments/assets/5ad97ab7-c995-4c8c-9c98-916c8819e38f)

---
 > we just need to store the totals somewhere.

Where would that be then, Supabase DB for this plugin or other storage solution?

> For example since you just checked today, and if I start a task tomorrow, it will only check my last one day of commits which should be a relatively tiny amount.

So we'd check commits from the beginning of time on the first run. We'd save the final output and the date we last ran a check. Then any subsequent usage of `/start` would parse commits across all repos from the last check date and it would update the storage with these new stats?

**Comment by 0x4007** [comment]
Ideally git based storage but for now I guess Supabase or whatever works.

**Comment by Keyrxng** [comment]
> maybe a separate app can handle that

Not ideal as partner' will need to create an additional app just to handle rate limits from a single plugin wouldn't they? or would it be our token that's used across all partners?

> The first run will be the heaviest,

So if we are looking at a 5 minute delay on the first run, then really we shouldn't assign the task right away or we are assigning right away?

> We also eventually should read private repos as well by authenticating through ubiquity-os

This doesn't work with using an additional app unless it too had private access. Wouldn't this plugin need the `APP_ID` and `APP_PRIVATE_KEY` to be able to do that, which is a no-no for plugins? Or use OAuth which isn't feasible in this context?

> I just realized with this plugin enabled we should reply with...comment before assigning. If they pass, then assign. If not, then edit the message saying that they require more experience. Perhaps something like

As I understand this comment ^

- if this plugin is not installed, assignment happens as normal
- if this plugin is installed, we comment first do our checks and then run `/start`

**Comment by Keyrxng** [comment]
> if this plugin is not installed, assignment happens as normal
if this plugin is installed, we comment first do our checks and then run /start

These two plugins were decoupled and run async as two separate plugins in the chain meaning they should both receive the payload at the same time. `/start` is going to finish first 9/10 so in order for us to avoid assigning them if this plugin is active what can we do?

Well, as two separate plugins that receive the same payload it's possible that this plugin forwards the same payload onto the `/start` worker, which is beneficial in that if we can stop `/start` from running via the command and instead have it run when it receives a payload directly from this plugin. Then it's sort of like it's running async.

Otherwise the only way I can see it happening is if `/start` becomes dependant on this plugin forwarding it's results via the kernel by using a chained config entry containing `xp-guard > command-start`

**Comment by Keyrxng** [comment]
> We also eventually should read private repos as well

I linked https://gitroll.io/ before and we could do something similar

- create a UI that the user logs in to via GitHub OAuth
- we index all of their repos with their own token and we store that, which this plugin would fetch from.

pros:

- we use the contributor's rate limit
- we have private repo access
- the user runs it whenever they want to to update their own stats (likely following a rejection)
- Improves the execution speed of this plugin and removes `getCommit` calls
- allows this to remain async without affecting `/start`

cons:

- requires more infra
- requires we modify our `/start` flow such that if an XP entry is not found for a contributor we must refer them to the UI in order to scan

**Comment by Keyrxng** [comment]
removing my assignment until the spec is made clearer

**Comment by 0x4007** [comment]
> removing my assignment until the spec is made clearer 

@rndquu rfc

**Comment by rndquu** [comment]
> > removing my assignment until the spec is made clearer
> 
> @rndquu rfc

We need a dead simple "dev experience checker" as described [here](https://github.com/ubiquity-os/plugins-wishlist/issues/26#issuecomment-2258827285)

**Comment by 0x4007** [comment]
Percentages are the wrong approach. The commit language percentage data is being abused to attempt to represent experience in a specific language. This is not accurate. 

My approach with the amount of commits with the different languages is far more accurate to represent experience.

**Comment by Keyrxng** [comment]
1. A user can easily push a ton of <lang> commits to any repo, spoofing commit content is also very easy.
2. Your approach requires we call into each commit to get all changes, 1 call per commit which is not scalable and I showed why. I forget the figures and QA I provided exactly but we are talking 1000s of calls per contributor and at least 3+ min response time (closer to 5 if I remember right just for me) so a team scenario is likely not feasible in most scenarios with the primary rate limit to consider.
3. Our rate limit will get destroyed, so we must create a new app/token just to be able to run your version.

---

> We need a dead simple "dev experience checker" as described https://github.com/ubiquity-os/plugins-wishlist/issues/26#issuecomment-2258827285

Perhaps it makes sense for you to just respec this task @0x4007. I feel this was part of my removal the beginnings of it because it seemingly became annoying with the back and forth.

I met @rndquu's spec and he has referred back a comment inline with my initial approach. This was ready to merge months ago according to the written spec and task author comments. 

Maybe it's best someone new works on it rather than myself but I'll still have both builds somewhere hopefully so might save time.
